### PR TITLE
Fix `distrib_type` not being set when Plugin instances being passed to Trainer

### LINKED ISSUE #10169: DDP on SLURM doesn't work when  `TrainingTypePlugin` is passed as strategy to `Trainer`
## 🐛 Bug
When passing `TrainingTypePlugin` as strategy to `Trainer`, as opposed to passing a string, DDP doesn't work on SLURM.

From a quick look, it seems the issue is that `self._distrib_type` is not set, causing properties of `AcceleratorConnector` to return a wrong value: `use_ddp` returns false instead of true which makes `_configure_slurm_ddp` do nothing.

Code for reproducing:
```
import os

import torch
from torch.utils.data import DataLoader, Dataset

from pytorch_lightning import LightningModule, Trainer
from pytorch_lightning.plugins import DDPPlugin
from pytorch_lightning.plugins.environments import SLURMEnvironment


class RandomDataset(Dataset):
    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len


class BoringModel(LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)

    def forward(self, x):
        return self.layer(x)

    def training_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("train_loss", loss)
        return {"loss": loss}

    def validation_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("valid_loss", loss)

    def test_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("test_loss", loss)

    def configure_optimizers(self):
        return torch.optim.SGD(self.layer.parameters(), lr=0.1)


def run():
    train_data = DataLoader(RandomDataset(32, 64), batch_size=2)
    val_data = DataLoader(RandomDataset(32, 64), batch_size=2)
    test_data = DataLoader(RandomDataset(32, 64), batch_size=2)

    model = BoringModel()

    # Choose one of these:
    strategy = DDPPlugin()
    # strategy = "ddp"

    trainer = Trainer(
        default_root_dir=os.getcwd(),
        limit_train_batches=1,
        limit_val_batches=1,
        num_sanity_val_steps=0,
        max_epochs=1,
        enable_model_summary=False,
        gpus=2,
        strategy=strategy,
    )
    information = (
        f"PID {os.getpid()}: Trainer strategy: {trainer._accelerator_connector.strategy}"
        + "\n"
        + f"PID {os.getpid()}: Is the trainer using SLURMEnvironment? {isinstance(trainer._accelerator_connector.cluster_environment, SLURMEnvironment)}"
    )
    print(information + "\n")

    # Launching these will result in two separated trainings when strategy == DDPPlugin():
    trainer.fit(model, train_dataloaders=train_data, val_dataloaders=val_data)
    trainer.test(model, dataloaders=test_data)


if __name__ == "__main__":
    run()
```

### PR DESCRIPTION
**Pull request by kaushikb11**
## What does this PR do?

Fixes `distrib_type` not being set when Plugin instances being passed to Trainer.

Also, fixes #10169 

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/plugins/training_type/ddp.py [hunk 1]
```diff
@@ -83,7 +84,7 @@ class DDPPlugin(ParallelPlugin):
     devices (e.g. GPU) per node. It is very similar to how :mod:`torch.distributed.launch` launches processes.
     """
 
-    distributed_backend = "ddp"
+    distributed_backend = DistributedType.DDP
```
[the remark this answers is not in the document]
Comment (kaushikb11) [reply pytorch_lightning/plugins/training_type/ddp.py:87]: Then, the `distributed_backend` would be `DistributedType.DDP`

**Review by justusschock — APPROVED** [review]

File: pytorch_lightning/trainer/connectors/accelerator_connector.py [hunk 2]
```diff
@@ -164,9 +165,13 @@ def __init__(
             self._set_training_type_plugin()
         else:
             self.set_distributed_mode()
-        self._configure_slurm_ddp()
 
         self.handle_given_plugins()
+        self._set_distrib_type_if_training_type_plugin_passed()
+
+        self._configure_slurm_ddp()
+        self.select_cluster_environment()
```
Comment (carmocca) [inline pytorch_lightning/trainer/connectors/accelerator_connector.py:173]: This function does not set state but instead returns an object, what does this line do then?

Comment (kaushikb11) [reply pytorch_lightning/trainer/connectors/accelerator_connector.py]: Resolved!

File: pytorch_lightning/plugins/training_type/ddp.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (four4fish) [reply pytorch_lightning/plugins/training_type/ddp.py:87]: That sounds misleading.

[the remark this answers is not in the document]
Comment (kaushikb11) [reply pytorch_lightning/plugins/training_type/ddp.py:87]: If you are subclassing `DDPPlugin` for `CustomDDPPlugin`, and the attribute `distributed_backend` is "ddp". It doesn't sound misleading to me. You could always override the attribute, it will update the `_distrib_type`. 

> CustomDDPPlugin won't exist in DistributedType.

If you are passing Plugin instances to Trainer, we don't check for `DistributedType`.

You could also check out [Plugins registry](https://pytorch-lightning.readthedocs.io/en/latest/advanced/plugins_registry.html) to register custom plugins.

**Review by rohitgr7 — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
LGMT !
